File: src/main/java/AoC2015_02.java
import static java.util.stream.Collectors.toList;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import lombok.RequiredArgsConstructor;
import lombok.ToString;

public final class AoC2015_02 extends SolutionBase<List<Present>, Integer, Integer> {

    private AoC2015_02(final boolean debug) {
        super(debug);
    }

    public static AoC2015_02 create() {
        return new AoC2015_02(false);
    }

    public static AoC2015_02 createDebug() {
        return new AoC2015_02(true);
    }

    @Override
    protected List<Present> parseInput(final List<String> inputs) {
        return inputs.stream().map(Present::fromInput).collect(toList());
    }

    @Override
    public Integer solvePart1(final List<Present> input) {
        return input.stream()
                .mapToInt(Present::calculateRequiredArea)
                .sum();
    }

    @Override
    public Integer solvePart2(final List<Present> input) {
        return input.stream()
                .mapToInt(Present::calculateRequiredLength)
                .sum();
    }

    @Override
    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "58"),
        @Sample(method = "part1", input = TEST2, expected = "43"),
        @Sample(method = "part2", input = TEST1, expected = "34"),
        @Sample(method = "part2", input = TEST2, expected = "14"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2015_02.create().run();
    }

    private static final String TEST1 = "2x3x4";
    private static final String TEST2 = "1x1x10";
}
    
@RequiredArgsConstructor
@ToString
final class Present {
    private final Integer length;
    private final Integer width;
    private final Integer height;
    
    public static Present fromInput(final String s) {
        final int[] sp = Stream.of(s.split("x")).mapToInt(Integer::valueOf).toArray();
        return new Present(sp[0], sp[1], sp[2]);
    }
    
    public int calculateRequiredArea() {
        final int[] sides = new int[] {
                2 * this.length * this.width,
                2 * this.width * this.height,
                2 * this.height * this.length};
        return Arrays.stream(sides).sum() + Arrays.stream(sides).min().getAsInt() / 2;
    }

    public int calculateRequiredLength() {
        final int[] circumferences = new int[] {
                2 * (this.length + this.width),
                2 * (this.width + this.height),
                2 * (this.height + this.length)};
        return Arrays.stream(circumferences).min().getAsInt()
                + this.length * this.width * this.height;
    }
}

File: src/main/cpp/2015/02/AoC2015_02.cpp
#include <assert.h>

#include <algorithm>
#include <climits>
#include <functional>
#include <numeric>
#include <string>
#include <vector>

#include "../../aoc/aoc.hpp"
#include "../../aocd/aocd.hpp"

using namespace std;

class Present {
   public:
    Present(const int length, const int width, const int height)
        : length(length), width(width), height(height) {}
    const int length;
    const int width;
    const int height;
};

vector<Present> parse(const vector<string>& input) {
    vector<Present> presents;
    for (const string& line : input) {
        const vector<string>& splits = aoc::split(line, "x");
        presents.push_back(
            Present(stoi(splits[0]), stoi(splits[1]), stoi(splits[2])));
    }
    return presents;
}

int calculateRequiredArea(const Present& present) {
    const vector<int> sides = {2 * present.length * present.width,
                               2 * present.width * present.height,
                               2 * present.height * present.length};
    int sum = 0, minimum = INT_MAX;
    for (int side : sides) {
        sum += side;
        minimum = min(side, minimum);
    }
    return sum + minimum / 2;
}

int calculateRequiredLength(const Present& present) {
    const vector<int> circumferences = {2 * (present.length + present.width),
                                        2 * (present.width + present.height),
                                        2 * (present.height + present.length)};
    return *min_element(circumferences.begin(), circumferences.end()) +
           present.length * present.width * present.height;
}

int solve(const vector<string>& input, const function<int(Present)>& strategy) {
    const vector<Present>& presents = parse(input);
    return accumulate(
        presents.begin(), presents.end(), 0,
        [&strategy](const int a, const Present& b) { return a + strategy(b); });
}

int part1(const vector<string>& input) {
    return solve(input, calculateRequiredArea);
}

int part2(const vector<string>& input) {
    return solve(input, calculateRequiredLength);
}

const vector<string> TEST1 = {"2x3x4"};
const vector<string> TEST2 = {"1x1x10"};

void samples() {
    assert(part1(TEST1) == 58);
    assert(part1(TEST2) == 43);
    assert(part2(TEST1) == 34);
    assert(part2(TEST2) == 14);
}

MAIN(2015, 2)
